src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -15,14 +15,57 @@
         task->reduce_scatter_kn.phase = _phase;                                \
     } while (0)
 
+static inline void get_sbuf_rbuf(ucc_knomial_pattern_t *p, ucc_coll_args_t *args,
+                                 void *scratch, size_t block_count,
+                                 void **sbuf, void **rbuf)
+{
+    uint8_t    node_type = p->node_type;
+    size_t     dt_size   = ucc_dt_size(args->dst.info.datatype);
+    ucc_rank_t rank      = p->rank;
+    ptrdiff_t  local_seg_offset;
+    size_t     local_seg_count;
+    size_t     offset;
+
+    if (ucc_knomial_pattern_loop_first_iteration(p)) {
+        *sbuf = (KN_NODE_PROXY == node_type || UCC_IS_INPLACE(*args))
+                 ? args->dst.info.buffer: args->src.info.buffer;
+        *rbuf = scratch;
+    } else if (ucc_knomial_pattern_loop_last_iteration(p)) {
+        ucc_sra_kn_get_offset_and_seglen(p->count, dt_size, rank, p->size,
+                                         p->radix, &local_seg_offset,
+                                         &local_seg_count);
+        /* for the last iteration of knomial algorithm, data is received
+           to the scratch buffer with different offset:
+              - if segment intersects with previous block offset is equal
+                to that block count
+              - otherwise offset is equal to that block count plus offset
+                required to make futher reduction strided
+            the goal is to save memcpy in the last step and do reduction to
+            correct destination directly
+        */
+        local_seg_offset /= dt_size;
+        if (local_seg_offset <= block_count || local_seg_count == 0) {
+            *sbuf = scratch;
+            *rbuf = PTR_OFFSET(*sbuf, block_count * dt_size);
+        } else {
+            offset = (local_seg_offset - block_count) % local_seg_count;
+            *sbuf = scratch;
+            *rbuf = PTR_OFFSET(*sbuf, (block_count + offset) * dt_size);
```

Comment:
looks good to me. just one question: if scratch is actual user rbuf (no extra + no_inplace case) do we always have enough space after "block_count" to recv segement. I think so but haven't proved on paper. Maybe add assert, like:
block_ount + offset + local_seg_count <= count.

Reply:
I think I checked all the corner cases and in every case where rbuf is used we have enough space. I will merge this in, let's see if any tests will fail until next release